vstools/exceptions/generic.py

```diff
@@ -153,84 +153,122 @@ def __init__(
         super().__init__(message, func, subsampling=subsampling, **kwargs)
 
 
-class FormatsMismatchError(CustomValueError):
+class MismatchError(CustomValueError):
+    """Raised when there's a mismatch between two or more values."""
+
+
+class FormatsMismatchError(MismatchError):
     """Raised when clips with different formats are given."""
 
     def __init__(
-        self, func: FuncExceptT, message: SupportsString = 'The format of both clips must be equal!',
+        self, clip_a: vs.RawNode, clip_b: vs.RawNode, func: FuncExceptT,
+        message: SupportsString = 'The format of both clips must be equal!',
+        reason: Any = '{a_format} != {b_format}',
         **kwargs: Any
     ) -> None:
-        super().__init__(message, func, **kwargs)
+        super().__init__(
+            message, func, reason, a_format=clip_a.format, b_format=clip_b.format, **kwargs
+        )
 
 
 class FormatsRefClipMismatchError(FormatsMismatchError):
     """Raised when a ref clip and the main clip have different formats"""
 
     def __init__(
-        self, func: FuncExceptT, message: SupportsString = 'The format of ref and main clip must be equal!',
+        self, clip: vs.RawNode, ref: vs.RawNode, func: FuncExceptT,
+        message: SupportsString = 'The format of ref and main clip must be equal!',
+        reason: Any = '{clip_format} != {ref_format}',
         **kwargs: Any
     ) -> None:
-        super().__init__(func, message, **kwargs)
+        super().__init__(
+            message, func, reason, clip_format=clip.format, ref_format=ref.format, **kwargs
+        )
 
 
-class ResolutionsMismatchError(CustomValueError):
+class ResolutionsMismatchError(MismatchError):
     """Raised when clips with different resolutions are given."""
 
     def __init__(
-        self, func: FuncExceptT, message: SupportsString = 'The resolution of both clips must be equal!',
+        self, clip_a: vs.RawNode, clip_b: vs.RawNode, func: FuncExceptT,
+        message: SupportsString = 'The resolution of both clips must be equal!',
+        reason: Any = '{a_width}x{a_height} != {b_width}x{b_height}',
```

Comment:
Would it make sense to extract this into `Resolution.__str__` for you?
Then passing `{a_res} != {b_res}` and `a_res=Resolution(clip_a.width, clip_a.height)`.

Reply:
Then we can accept `Resolution | vs.VideoNode`

Reply:
Might be a good idea. Alternatively, would it make sense to have this inside of VS itself? Maybe not, but at the same time?